File: JSCodes/src/communication/QuantumEncryption.js
const { webcrypto } = require('crypto');
const { ErrorHandler, ErrorSeverity } = require('../utils/ErrorHandling');

class QuantumEncryption {
    constructor(keyLength = 256) {
        this.keyLength = keyLength;
        this.crypto = webcrypto;
    }

    async generateQuantumKey() {
        try {
            // Generate random bytes for quantum-like key
            const key = await this.crypto.getRandomValues(new Uint8Array(this.keyLength / 8));
            return key;
        } catch (error) {
            throw ErrorHandler.createError(
                `Quantum key generation failed: ${error}`,
                ErrorSeverity.ERROR,
                'QUANTUM_ENCRYPTION'
            );
        }
    }

    async encryptMessage(message, key) {
        try {
            const messageBytes = new TextEncoder().encode(message);
            const encrypted = new Uint8Array(messageBytes.length);

            // XOR encryption with quantum key
            for (let i = 0; i < messageBytes.length; i++) {
                encrypted[i] = messageBytes[i] ^ key[i % key.length];
            }

            return encrypted;
        } catch (error) {
            throw ErrorHandler.createError(
                `Encryption failed: ${error}`,
                ErrorSeverity.ERROR,
                'QUANTUM_ENCRYPTION'
            );
        }
    }

    async decryptMessage(encrypted, key) {
        try {
            const decrypted = new Uint8Array(encrypted.length);

            // XOR decryption with quantum key
            for (let i = 0; i < encrypted.length; i++) {
                decrypted[i] = encrypted[i] ^ key[i % key.length];
            }

            return new TextDecoder().decode(decrypted);
        } catch (error) {
            throw ErrorHandler.createError(
                `Decryption failed: ${error}`,
                ErrorSeverity.ERROR,
                'QUANTUM_ENCRYPTION'
            );
        }
    }

    async createEntangledPair() {
        try {
            // Simulate quantum entanglement with correlated random values
            const pair1 = await this.generateQuantumKey();
            const pair2 = pair1.slice(0); // Create copy with same values

            return [pair1, pair2];
        } catch (error) {
            throw ErrorHandler.createError(
                `Entangled pair creation failed: ${error}`,
                ErrorSeverity.ERROR,
                'QUANTUM_ENCRYPTION'
            );
        }
    }
}

module.exports = QuantumEncryption; 

File: PythonCodes/src/communication/quantum_encryption.py
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister, execute, Aer
from qiskit.circuit.library import QFT
import numpy as np
from typing import Tuple, List
from ..utils.error_handling import KAPESITError, ErrorSeverity

class QuantumEncryption:
    def __init__(self, key_length: int = 256):
        self.key_length = key_length
        self.simulator = Aer.get_backend('qasm_simulator')
        
    def generate_quantum_key(self) -> List[int]:
        qr = QuantumRegister(self.key_length)
        cr = ClassicalRegister(self.key_length)
        circuit = QuantumCircuit(qr, cr)
        
        # Apply Hadamard gates to create superposition
        for i in range(self.key_length):
            circuit.h(qr[i])
            
        # Measure qubits
        circuit.measure(qr, cr)
        
        # Execute circuit
        job = execute(circuit, self.simulator, shots=1)
        result = job.result()
        counts = result.get_counts(circuit)
        
        # Convert measurement to binary key
        key = [int(bit) for bit in list(counts.keys())[0]]
        return key
    
    def encrypt_message(self, message: str, key: List[int]) -> bytes:
        try:
            message_bytes = message.encode('utf-8')
            encrypted = bytearray()
            
            for i, byte in enumerate(message_bytes):
                key_byte = key[i % len(key)]
                encrypted.append(byte ^ key_byte)
                
            return bytes(encrypted)
        except Exception as e:
            raise KAPESITError(
                f"Encryption failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="QUANTUM_ENCRYPTION"
            )
    
    def decrypt_message(self, encrypted: bytes, key: List[int]) -> str:
        try:
            decrypted = bytearray()
            
            for i, byte in enumerate(encrypted):
                key_byte = key[i % len(key)]
                decrypted.append(byte ^ key_byte)
                
            return decrypted.decode('utf-8')
        except Exception as e:
            raise KAPESITError(
                f"Decryption failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="QUANTUM_ENCRYPTION"
            )
    
    def create_entangled_pair(self) -> Tuple[QuantumCircuit, QuantumCircuit]:
        qr = QuantumRegister(2)
        circuit = QuantumCircuit(qr)
        
        # Create Bell state
        circuit.h(qr[0])
        circuit.cx(qr[0], qr[1])
        
        return circuit, circuit.copy() 